File: ruby/day18.rb
require "set"

def main
  lines = $stdin.readlines.map { |l| eval(l) }
  res = lines.inject { |a, b| reduce([a,b]) }
  p res
  puts mag(res)
  #p result: reduce([[[[[9,8],1],2],3],4])
  #puts "------"
  #p result: reduce([7,[6,[5,[4,[3,2]]]]])
  #puts "------"
  #p result: reduce([[3,[2,[1,[7,3]]]],[6,[5,[4,[3,2]]]]])
  #puts "------"
  #p result: reduce([[[[[4,3],4],4],[7,[[8,4],9]]],[1,1]])
end

def mag(sn)
  case sn
  when Array
    a, b = sn
    3*mag(a) + 2*mag(b)
  else
    sn
  end
end

def reduce(sn)
  p input: sn
  sn = convert(sn)
  loop do
    next if explode(sn)
    next if split(sn)
    return unconvert(sn)
  end
end

def convert(sn, depth: 0, res: [])
  case sn
  when Array
    convert(sn[0], depth: depth+1, res: res)
    convert(sn[1], depth: depth+1, res: res)
  else
    res << [sn, depth]
  end
  res
end

def unconvert(sn)
  p unconvert: sn
  res = [[]]
  while n = sn.shift
    #p res: res, n: n
    while n[1] < res.size
      loop do
        a = res.pop
        res.last.push(a)
        break unless res.last.size == 2
      end
    end
    while n[1] > res.size
      res.push([])
    end
    res.last.push(n[0])
    while res.last.size == 2 && res.size > 1
      a = res.pop
      res.last.push(a)
    end
  end
  until res.size == 1
    a = res.pop
    res.last.push(a)
  end
  res.pop
end

def explode(sn)
  (0..sn.size-1).each do |i|
    if sn[i][1] > 4 && sn[i+1][1] == sn[i][1]
      p1 = sn[i]
      p2 = sn.delete_at(i+1)
      sn[i] = [0, p1[1]-1]
      if i > 0
        n1 = sn[i-1]
        n1[0] += p1[0]
      end
      if n2 = sn[i+1]
        n2[0] += p2[0]
      end
      return true
    end
  end
  false
end

def split(sn)
  i = sn.index { |x,d| x >= 10 }
  return false if i.nil?
  x, d = sn[i]
  xa = x/2
  xb = x-xa
  sn[i] = [xa, d+1]
  sn.insert(i+1, [xb, d+1])
  true
end

main

File: src/day18.rs
use super::common::read_lines::read_lines;
use std::io::Read;

type Value = usize;

#[derive(Clone)]
enum SN {
    Leaf(Value),
    Pair(Box<SN>, Box<SN>),
}

pub fn run<R: Read>(r: R) {
    let trees = read_lines(r)
        .map(|line| parse_tree(&line).0)
        .collect::<Vec<SN>>();
    let part1 = trees
        .iter()
        .cloned()
        .reduce(|a, b| reduce(sum(a, b)))
        .unwrap();
    println!("part 1: {}", mag(&part1));
}

fn parse_tree(line: &str) -> (SN, &str) {
    if c0(line) == '[' {
        let (l, line) = parse_tree(&line[1..]);
        assert_eq!(',', c0(line));
        let (r, line) = parse_tree(&line[1..]);
        assert_eq!(']', c0(line));
        (pair(l, r), &line[1..])
    } else {
        let i = line.find(|c| c == ']' || c == ',').unwrap();
        let val = line[0..i].parse().unwrap();
        let line = &line[i..];
        (SN::Leaf(val), line)
    }
}

fn pair(l: SN, r: SN) -> SN {
    SN::Pair(Box::new(l), Box::new(r))
}

fn c0(line: &str) -> char {
    line.chars().next().unwrap()
}

fn sum(l: SN, r: SN) -> SN {
    pair(l, r)
}

fn reduce(mut root: SN) -> SN {
    loop {
        println!("{:?}", &root);
        if let Some(new_root) = finish_exploding(explode(&root, 0)) {
            root = new_root;
            continue;
        }
        if let Some(new_root) = split(&root) {
            root = new_root;
            continue;
        }
        return root;
    }
}

fn split(node: &SN) -> Option<SN> {
    match node {
        SN::Pair(l, r) => match split(l) {
            Some(nl) => Some(pair(nl, unbox(r))),
            None => match split(r) {
                Some(nr) => Some(pair(unbox(l), nr)),
                None => None,
            },
        },
        SN::Leaf(v) if *v > 9 => {
            let l = *v / 2;
            let r = *v - l;
            Some(pair(SN::Leaf(l), SN::Leaf(r)))
        }
        SN::Leaf(_) => None,
    }
}

fn finish_exploding(ex: Explode) -> Option<SN> {
    match ex {
        Explode::Exploded(_, _) => panic!("exploded at root?!?"),
        Explode::Nothing(_) => None,
        Explode::ExplodedLeft(sn, _) => Some(sn),
        Explode::ExplodedRight(sn, _) => Some(sn),
        Explode::ExplodedDone(sn) => Some(sn),
    }
}

enum Explode {
    Nothing(SN),
    Exploded(Value, Value),
    ExplodedLeft(SN, Value),
    ExplodedRight(SN, Value),
    ExplodedDone(SN),
}

fn explode(node: &SN, depth: usize) -> Explode {
    match node {
        SN::Pair(l, r) if l.is_leaf() && r.is_leaf() => {
            if depth > 3 {
                println!("explode!");
                Explode::Exploded(l.value().unwrap(), r.value().unwrap())
            } else {
                Explode::Nothing(pair(unbox(l), unbox(r)))
            }
        }
        SN::Pair(l, r) => {
            match explode(l, depth + 1) {
                Explode::Exploded(el, er) => {
                    // parent needs to add el to its right-most value.
                    Explode::ExplodedLeft(pair(SN::Leaf(0), add_to_left(r, er)), el)
                }
                Explode::ExplodedLeft(nl, el) => {
                    // keep passing el up, parent needs to deal with it.
                    Explode::ExplodedLeft(pair(nl, unbox(r)), el)
                }
                Explode::ExplodedRight(nl, er) => {
                    // we have a value that can accept it!
                    Explode::ExplodedDone(pair(nl, add_to_left(r, er)))
                }
                Explode::ExplodedDone(nl) => Explode::ExplodedDone(pair(nl, unbox(r))),
                Explode::Nothing(nl) => match explode(r, depth + 1) {
                    Explode::Exploded(el, er) => {
                        Explode::ExplodedRight(pair(add_to_right(&nl, el), SN::Leaf(0)), er)
                    }
                    Explode::ExplodedLeft(nr, el) => {
                        Explode::ExplodedDone(pair(add_to_right(&nl, el), nr))
                    }
                    Explode::ExplodedRight(nr, er) => Explode::ExplodedRight(pair(nl, nr), er),
                    Explode::ExplodedDone(nr) => Explode::ExplodedDone(pair(nl, nr)),
                    Explode::Nothing(nr) => Explode::Nothing(pair(nl, nr)),
                },
            }
        }
        n => Explode::Nothing(n.clone()),
    }
}

fn add_to_right(sn: &SN, val: Value) -> SN {
    match sn {
        SN::Pair(l, r) => pair(unbox(l), add_to_right(r, val)),
        SN::Leaf(v) => SN::Leaf(v + val),
    }
}

fn add_to_left(sn: &SN, val: Value) -> SN {
    match sn {
        SN::Pair(l, r) => pair(add_to_left(l, val), unbox(r)),
        SN::Leaf(v) => SN::Leaf(v + val),
    }
}

fn unbox(n: &SN) -> SN {
    n.clone()
}

fn mag(n: &SN) -> Value {
    match n {
        SN::Leaf(x) => *x,
        SN::Pair(l, r) => 3 * mag(l) + 2 * mag(r),
    }
}

impl SN {
    pub fn is_leaf(&self) -> bool {
        match self {
            SN::Leaf(_) => true,
            _ => false,
        }
    }

    pub fn value(&self) -> Option<Value> {
        match self {
            SN::Leaf(v) => Some(*v),
            _ => None,
        }
    }
}

impl std::fmt::Debug for SN {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            SN::Pair(l, r) => write!(f, "[{:?},{:?}]", l, r),
            SN::Leaf(v) => write!(f, "{}", v),
        }
    }
}
